src/java.desktop/macosx/classes/com/apple/laf/AquaImageFactory.java

```diff
@@ -231,10 +232,24 @@ protected IconUIResource getInstance() {
     @SuppressWarnings("serial") // Superclass is not serializable across versions
     static class InvertableImageIcon extends ImageIcon implements InvertableIcon, UIResource {
         Icon invertedImage;
+        Icon disabledIcon;
         public InvertableImageIcon(final Image image) {
             super(image);
         }
 
+        @Override
+        public void paintIcon(Component c, Graphics g, int x, int y) {
+            if (!c.isEnabled()) {
+                if (disabledIcon == null) {
+                    disabledIcon = new ImageIconUIResource(GrayFilter.
+                            createDisabledImage(((ImageIcon)this).getImage()));
+                }
+                disabledIcon.paintIcon(c, g, x, y);
+            } else {
+                super.paintIcon(c, g, x, y);
+            }
+        }
+
         @Override
         public Icon getInvertedIcon() {
```

Comment:
When we draw result of this method? Is it possible that we use the getInvertedIcon for disabled menus as well?

Reply:
It is called from AquaMenuPainter#405 and 378
if (c instanceof JMenu && (model.isArmed() || model.isSelected()) && arrowIcon instanceof InvertableIcon) {
            ((InvertableIcon)arrowIcon).getInvertedIcon().paintIcon(c, g, arrowIconRect.x, arrowIconRect.y);
        }
I could not find at what point arrowIcon or checkIcon would be instance of InvertableIcon...
I guess we can take one fix at a time since this normal icon disable issue itself is there from macox port origin 
and fix it later if any issue comes for InvertableIcon...

Reply:
> I could not find at what point arrowIcon or checkIcon would be an instance of InvertableIcon...

All InvertableImageIcon objects are InvertableIcon since it implements that interface, so that part is always true, and if it possible to make "(c instanceof JMenu && (model.isArmed() || model.isSelected()"== true for the disabled components then "getInvertedIcon" should return disabled icon.

Reply:
Arrow icon InvertedIcon is handled by testing with setSelected(true) and setEnabled(false)
but
for checkIcon, it checks for
```
if (item.isArmed() && checkIcon instanceof InvertableIcon) {
            ((InvertableIcon)checkIcon).getInvertedIcon().paintIcon(item, g, checkIconRect.x, checkIconRect.y);
        } 

```
but even if we do setArmed(true) it is set to false here
https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/JMenuItem.java#L312
```
if (!b && !UIManager.getBoolean("MenuItem.disabledAreNavigable")) {
            setArmed(false);
        }
```
and MenuItem.disabledAreNavigable is not defined for Aqua so I guess it's false so InvertedIcon is not called for checkIcon